### PR TITLE
{AppService} `az webapp`: Rename container related parameters and add parameters `--container-image-name` and `--container-registry-url` into commands `az webapp create` and `az webapp deployment slot create` to deprecate `--deployment-container-image-name`

### PR DESCRIPTION
**Pull request by mstxsun**
**Related command**
`az webapp create`, `az webapp deployment slot create`, `az webapp config container set`

**Description**
Remove docker prefix and rename container webapp related parameters
`--docker-registry-server-url`
`--docker-custom-image-name`
`--docker-registry-server-user`
`--docker-registry-server-password`
to
`--container-registry-url`
`--container-image-name`
`--container-registry-user`
`--container-registry-password`
and adding parameters `--container-image-name` and `container-registry-url` into `az webapp create` and `az webapp deployment slot create` to deprecate parameter `--deployment-container-image-name`. However, we will keep `--deployment-container-image-name` and allow customer to still use it for now.

**Testing Guide**
1. Ran commands `az webapp create`, `az webapp deployment slot create`, `az webapp config container set` and check if those changes break them
2. Ran appservice tests locally

**History Notes**

[AppService] `az webapp config container set`: Remove docker prefix and rename container related parameters
[AppService] `az webapp create` and `az webapp deployment slot create`: Rename `--docker-registry-server-user` and `--docker-registry-server-password` to `--container-registry-user` and `--container-registry-password` and add parameters `--container-image-name` and `container-registry-url` to deprecate `--deployment-container-image-name`
[AppService] `az webapp create` and `az logicapp create`: Update help message

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
AppService

**Comment by mstxsun** [comment]
@jvano

File: src/azure-cli/azure/cli/command_modules/appservice/_params.py [hunk 1]
```diff
@@ -147,8 +147,12 @@ def load_arguments(self, _):
                    local_context_attribute=LocalContextAttribute(name='web_name', actions=[LocalContextAction.SET],
                                                                  scopes=['webapp', 'cupertino']))
         c.argument('startup_file', help="Linux only. The web's startup file")
-        c.argument('docker_registry_server_user', options_list=['--docker-registry-server-user', '-s'], help='the container registry server username')
-        c.argument('docker_registry_server_password', options_list=['--docker-registry-server-password', '-w'], help='The container registry server password. Required for private registries.')
+        c.argument('deployment_container_image_name', options_list=['--deployment-container-image-name', '-i'], help='Container image name from Docker Hub, e.g. publisher/image-name:tag', deprecate_info=c.deprecate(target='--deployment-container-image-name'))
```
Comment (jvano) [inline src/azure-cli/azure/cli/command_modules/appservice/_params.py:151]: We should not call our "Docker Hub" in any place. If you need to talk about a container image repository, we called them "container registry"

File: src/azure-cli/azure/cli/command_modules/appservice/_params.py [hunk 2]
```diff
@@ -838,6 +844,7 @@ def load_arguments(self, _):
         c.argument('name', arg_type=logicapp_name_arg_type)
 
     with self.argument_context('logicapp create') as c:
+        c.argument('deployment_container_image_name', options_list=['--deployment-container-image-name', '-i'], help='Container image name from Docker Hub, e.g. publisher/image-name:tag')
```
Comment (jvano) [inline src/azure-cli/azure/cli/command_modules/appservice/_params.py:880]: same here

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 3]
```diff
@@ -115,8 +116,22 @@ def create_webapp(cmd, resource_group_name, name, plan, runtime=None, startup_fi
...
+    if deployment_container_image_name and container_image_name:
+        raise MutuallyExclusiveArgumentError('usage error: --deployment-container-image-name | --container-image-name')
+    if container_registry_url and not container_image_name:
+        raise ArgumentUsageError("Please use --container-image-name to provide the image name")
 
-    docker_registry_server_url = parse_docker_image_name(deployment_container_image_name)
+    if container_registry_url:
+        container_registry_url = parse_container_registry_url(container_registry_url)
+    else:
+        container_registry_url = parse_docker_image_name(deployment_container_image_name)
```
Comment (jvano) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:131]: rename the methos to parse_container_image_name as we are trying to get rid of "docker" when possible.

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 4]
```diff
@@ -2066,7 +2114,7 @@ def create_functionapp_slot(cmd, resource_group_name, name, slot, configuration_
         update_slot_configuration_from_source(cmd, client, resource_group_name, name, slot, configuration_source,
                                               image, registry_password,
                                               registry_username,
-                                              docker_registry_server_url=docker_registry_server_url)
+                                              container_registry_url=docker_registry_server_url)
```
Comment (jvano) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:2455]: also check if we can rename docker_registry_server_url to container_registry_server_url

**Review by jvano — CHANGES_REQUESTED** [review]

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py (hunk 3 above)
Comment (mstxsun) [reply src/azure-cli/azure/cli/command_modules/appservice/custom.py:131]: Will do this renaming in another PR

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py (hunk 4 above)
Comment (mstxsun) [reply src/azure-cli/azure/cli/command_modules/appservice/custom.py:2455]: This docker_registry_server_url is inside functionapp, will have separate PR for them if renaming in functionapp is needed.

**Comment by yanzhudd** [comment]
since there are muliple changes in this PR, it would be good to separately list all of the changes in the History Notes section above

**Comment by mstxsun** [comment]
> since there are muliple changes in this PR, it would be good to separately list all of the changes in the History Notes section above

@yanzhudd thanks for your comment, I updated History Note and put separate records of my changes there. Please check it out, thanks

**Comment by yanzhudd** [comment]
Let me check with you if there is any breaking change in this PR?

File: src/azure-cli/azure/cli/command_modules/appservice/_params.py [hunk 5]
```diff
@@ -148,8 +148,12 @@ def load_arguments(self, _):
...
                                                                  scopes=['webapp', 'cupertino']))
         c.argument('startup_file', help="Linux only. The web's startup file")
-        c.argument('docker_registry_server_user', options_list=['--docker-registry-server-user', '-s'], help='the container registry server username')
-        c.argument('docker_registry_server_password', options_list=['--docker-registry-server-password', '-w'], help='The container registry server password. Required for private registries.')
+        c.argument('deployment_container_image_name', options_list=['--deployment-container-image-name', '-i'], help='Container image name from container registry, e.g. publisher/image-name:tag', deprecate_info=c.deprecate(target='--deployment-container-image-name'))
+        c.argument('container_registry_url', options_list=['--container-registry-url'], help='the container registry server url')
+        c.argument('container_image_name', options_list=['--container-image-name', '-c'],
+                   help='the container custom image name and optionally the tag name (e.g., <registry-name>/<image-name>:<tag>)')
```
Comment (yanzhudd) [inline src/azure-cli/azure/cli/command_modules/appservice/_params.py:154]:
```suggestion
        c.argument('container_registry_url', options_list=['--container-registry-url'], help='The container registry server url')
        c.argument('container_image_name', options_list=['--container-image-name', '-c'],
                   help='The container custom image name and optionally the tag name (e.g., <registry-name>/<image-name>:<tag>)')
```

File: src/azure-cli/azure/cli/command_modules/appservice/_params.py [hunk 6]
```diff
@@ -148,8 +148,12 @@ def load_arguments(self, _):
...
+        c.argument('container_registry_user', options_list=['--container-registry-user', '-s', c.deprecate(target='--docker-registry-server-user', redirect='--container-registry-user')], help='the container registry server username')
```
Comment (yanzhudd) [inline src/azure-cli/azure/cli/command_modules/appservice/_params.py:155]:
```suggestion
        c.argument('container_registry_user', options_list=['--container-registry-user', '-s', c.deprecate(target='--docker-registry-server-user', redirect='--container-registry-user')], help='The container registry server username')
```

File: src/azure-cli/azure/cli/command_modules/appservice/_params.py [hunk 7]
```diff
@@ -457,10 +460,13 @@ def load_arguments(self, _):
         c.argument('configuration_source',
                    help="source slot to clone configurations from. Use web app's name to refer to the production slot")
         c.argument('deployment_container_image_name', options_list=['--deployment-container-image-name', '-i'],
-                   help='Container image name, e.g. publisher/image-name:tag')
-        c.argument('docker_registry_server_password', options_list=['--docker-registry-server-password', '-w'],
+                   help='Container image name, e.g. publisher/image-name:tag', deprecate_info=c.deprecate(target='--deployment-container-image-name'))
+        c.argument('container_registry_url', options_list=['--container-registry-url', '-r'], help='the container registry server url')
```
Comment (yanzhudd) [inline src/azure-cli/azure/cli/command_modules/appservice/_params.py:464]:
```suggestion
        c.argument('container_registry_url', options_list=['--container-registry-url', '-r'], help='The container registry server url')
```

File: src/azure-cli/azure/cli/command_modules/appservice/_params.py [hunk 8]
```diff
@@ -457,10 +460,13 @@ def load_arguments(self, _):
...
+        c.argument('container_image_name', options_list=['--container-image-name', '-c'],
+                   help='the container custom image name and optionally the tag name (e.g., <registry-name>/<image-name>:<tag>)')
+        c.argument('container_registry_user', options_list=['--container-registry-user', '-u', c.deprecate(target='--docker-registry-server-user', redirect='--container-registry-user')], help='the container registry server username')
```
Comment (yanzhudd) [inline src/azure-cli/azure/cli/command_modules/appservice/_params.py:467]:
```suggestion
        c.argument('container_image_name', options_list=['--container-image-name', '-c'],
                   help='The container custom image name and optionally the tag name (e.g., <registry-name>/<image-name>:<tag>)')
        c.argument('container_registry_user', options_list=['--container-registry-user', '-u', c.deprecate(target='--docker-registry-server-user', redirect='--container-registry-user')], help='The container registry server username')
```

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 9]
```diff
@@ -205,10 +221,17 @@ def create_webapp(cmd, resource_group_name, name, plan, runtime=None, startup_fi
...
-        if not validate_container_app_create_options(runtime, deployment_container_image_name,
+        if not validate_container_app_create_options(runtime, container_image_name,
                                                      multicontainer_config_type, multicontainer_config_file):
-            raise ArgumentUsageError("usage error: --runtime | --deployment-container-image-name |"
-                                     " --multicontainer-config-type TYPE --multicontainer-config-file FILE")
+            if deployment_container_image_name:
+                raise ArgumentUsageError('Please specify both --multicontainer-config-type TYPE '
+                                         'and multicontainer-config-file FILE, '
+                                         'and only specify one out of --runtime, '
+                                         '--deployment-container-image-name and --multicontainer-config-type')
```
Comment (yanzhudd) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:230]:
```suggestion
                raise ArgumentUsageError('Please specify both --multicontainer-config-type TYPE '
                                         'and --multicontainer-config-file FILE, '
                                         'and only specify one out of --runtime, '
                                         '--deployment-container-image-name and --multicontainer-config-type')
```

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 10]
```diff
@@ -205,10 +221,17 @@ def create_webapp(cmd, resource_group_name, name, plan, runtime=None, startup_fi
...
+            raise ArgumentUsageError('Please specify both --multicontainer-config-type TYPE '
+                                     'and multicontainer-config-file FILE, '
+                                     'and only specify one out of --runtime, '
+                                     '--container-image-name and --multicontainer-config-type')
```
Comment (yanzhudd) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:234]:
```suggestion
            raise ArgumentUsageError('Please specify both --multicontainer-config-type TYPE '
                                     'and --multicontainer-config-file FILE, '
                                     'and only specify one out of --runtime, '
                                     '--container-image-name and --multicontainer-config-type')
```

**Comment by yanzhudd** [comment]
please rerun some related tests in live mode to check if it meets your expectation

**Comment by yanzhudd** [comment]
Please note that Azure CLI team is not responsible for code style and specification, but not business logic. It is better to get approval from service team for business logic as well.

**Comment by mstxsun** [comment]
> Let me check with you if there is any breaking change in this PR?

I believe there is no breaking change in this PR, we don't remove those old parameters when we added support on new parameters, so customer should still be able to do what they want to do in the old way.

**Review by jvano — APPROVED** [review]
Looks good!

**Review by yanzhudd — APPROVED** [review]

**Comment by yanzhudd** [comment]
Hi @mstxsun,
it seems that there are multiple deprecated parameters in this PR, and we are about to having a breaking change window next sprint. Let me check with you if you plan to remove these deprecated parameter in the breaking change window. If so or you plan to make any other breaking changes, please let us know. Thanks!

**Pull request merged**